src/routes/conversation/[id]/+server.ts

```diff
@@ -191,105 +184,25 @@ export async function POST({ request, fetch, locals, params, getClientAddress })
 				webSearchResults = await runWebSearch(conv, newPrompt, update);
 			}
 
-			// we can now build the prompt using the messages
-			const prompt = await buildPrompt({
-				messages,
-				model,
-				webSearch: webSearchResults,
-				preprompt: conv.preprompt ?? model.preprompt,
-				locals: locals,
-			});
-
-			// fetch the endpoint
-			const randomEndpoint = modelEndpoint(model);
-
-			let usedFetch = fetch;
+			messages[messages.length - 1].webSearch = webSearchResults;
 
-			if (randomEndpoint.host === "sagemaker") {
-				const aws = new AwsClient({
-					accessKeyId: randomEndpoint.accessKey,
-					secretAccessKey: randomEndpoint.secretKey,
-					sessionToken: randomEndpoint.sessionToken,
-					service: "sagemaker",
-				});
+			conv.messages = messages;
 
-				usedFetch = aws.fetch.bind(aws) as typeof fetch;
-			}
-
-			async function saveLast(generated_text: string) {
-				if (!conv) {
-					throw error(404, "Conversation not found");
-				}
-
-				const lastMessage = messages[messages.length - 1];
-
-				if (lastMessage) {
-					// We could also check if PUBLIC_ASSISTANT_MESSAGE_TOKEN is present and use it to slice the text
-					if (generated_text.startsWith(prompt)) {
-						generated_text = generated_text.slice(prompt.length);
-					}
-
-					generated_text = trimSuffix(
-						trimPrefix(generated_text, "<|startoftext|>"),
-						PUBLIC_SEP_TOKEN
-					).trimEnd();
-
-					// remove the stop tokens
-					for (const stop of [...(model?.parameters?.stop ?? []), "<|endoftext|>"]) {
-						if (generated_text.endsWith(stop)) {
-							generated_text = generated_text.slice(0, -stop.length).trimEnd();
-						}
-					}
-					lastMessage.content = generated_text;
-
-					await collections.conversations.updateOne(
-						{
-							_id: convId,
-						},
-						{
-							$set: {
-								messages,
-								title: conv.title,
-								updatedAt: new Date(),
-							},
-						}
-					);
-
-					update({
-						type: "finalAnswer",
-						text: generated_text,
-					});
-				}
-			}
-
-			const tokenStream = textGenerationStream(
-				{
-					parameters: {
-						...models.find((m) => m.id === conv.model)?.parameters,
-						return_full_text: false,
-					},
-					model: randomEndpoint.url,
-					inputs: prompt,
-					accessToken: randomEndpoint.host === "sagemaker" ? undefined : HF_ACCESS_TOKEN,
-				},
-				{
-					use_cache: false,
-					fetch: usedFetch,
-				}
-			);
+			const endpoint = await model.getEndpoint();
 
-			for await (const output of tokenStream) {
```

Comment:
Please consider refactor to move `await endpoint({ conversation: conv }))`  out of `ReadableStream()` and/or out of `Response(stream)`

In practice it's quite common `await endpoint()` throws, for example getting `409` from server because of hitting rate limiter from server, for configuration was wrong and getting `404`.

IMO it make little sense to start streaming with we don't know if the response return OK

for example if something throw an exception in start handler in `ReadableStream`:
```ts
const stream = new ReadableStream({
    async start(controller) {
        throw new Error("broken")
    }
});
return new Response(stream);
```

Client will just get an `Network` error, IMO we should only return `Response`, i.e., start streaming when external API are green.
```ts
// handle error before returning `new Response(stream)`.
try {
  const example = await endpoint({ conversation: conv }));
} catch (e) {
  throw error(500, e.message);
}
const stream = new ReadableStream({
    async start(controller) {
        for await (const output of example) {
            ...
        }
    }
});
return new Response(stream);
```

Reply:
I think that's a valid point but a bit more intense of a refactor than I think I want to do in this PR specifically. 

We currently need to be in the stream before creating the token stream. For example, when we do requests to the web search that happens before creating the token stream but we also send updates about that so we need to be streaming already...